File: test/data/cpp/test01/StateController.cpp
#include "StateController.h"
#include <string.h>
StateController::StateController(IControllee* controllee)
{
    _controllee = controllee;
    InstanceOfState1 = new StateState1(this, _controllee);
    InstanceOfState2 = new StateState2(this, _controllee);
    _currentState = InstanceOfState1;
    BaseState* stateList[] = {
        InstanceOfState1,
        InstanceOfState2
    };
    memcpy(_stateList, stateList, sizeof(BaseState*) * MaxNumberOfStateIDs);
}
StateController::~StateController()
{
    delete InstanceOfState1;
    delete InstanceOfState2;
}
int StateController::GetCurrentIdFromStateId(int parentStateId)
{
    if(parentStateId == -1)
    {
        return GetCurrentIdFromState(_currentState);
    }
    if(parentStateId < -1 || parentStateId >= MaxNumberOfStateIDs)
    {
        return -1;
    }
    return GetCurrentIdFromState(_stateList[parentStateId]->CurrentSubState());
}
int StateController::GetCurrentIdFromState(BaseState* state)
{
    if(state == 0)
    {
        return -1;
    }
    return state->GetStateID();
    }
void StateController::ResumeState(int parentStateId, int stateId)
{
    BaseState* state = 0;
    if(stateId >= 0 && stateId < MaxNumberOfStateIDs)
    {
        state = _stateList[stateId];
    }
    if(parentStateId == -1)
    {
        _currentState = state;
        return;
    }
    if(parentStateId < -1 || parentStateId >= MaxNumberOfStateIDs)
    {
        return;
    }
    _stateList[parentStateId]->ResumeSubState(state);
}
const char* StateController::StateName(int parentStateId)
{
    if(parentStateId >= 0 && parentStateId < MaxNumberOfStateIDs)
    {
        return _stateList[parentStateId]->GetStateName();
    }
    return "(None)";
}
bool StateController::TryTransitWithoutEvent()
{
    if(_currentState == 0)
    {
        return false;
    }
    BaseState* current = _currentState;
    _currentState = _currentState->TryTransitWithoutEvent();
    if(_currentState != 0)
    {
        _currentState = _currentState->OutlineState();
    }
    return (current != _currentState);
}
void StateController::TransitCommand1()
{
    if(_currentState != 0)
    {
        _currentState = _currentState->TransitCommand1();
        if(_currentState != 0)
        {
            _currentState = _currentState->OutlineState();
        }
    } else {
        _controllee->OverTransition("Command1");
    }
}
void StateController::TransitCommand2()
{
    if(_currentState != 0)
    {
        _currentState = _currentState->TransitCommand2();
        if(_currentState != 0)
        {
            _currentState = _currentState->OutlineState();
        }
    } else {
        _controllee->OverTransition("Command2");
    }
}
void StateController::TransitCommand3()
{
    if(_currentState != 0)
    {
        _currentState = _currentState->TransitCommand3();
        if(_currentState != 0)
        {
            _currentState = _currentState->OutlineState();
        }
    } else {
        _controllee->OverTransition("Command3");
    }
}
const char* StateController::GetCurrentStateName()
{
    if(_currentState == 0)
    {
        return "(end)";
    }
    else
    {
        return _currentState->GetStateName();
    }
}

File: test/data/typescript/test06/StateController.ts
import {IControllee} from "./IControllee"
import {BaseState} from "./BaseState"
import {StateProcessData} from "./StateProcessData"
import {StateState1} from "./StateState1"
import {StateState2} from "./StateState2"
import {StateState3} from "./StateState3"
import {StateState4} from "./StateState4"
import {Statelong1} from "./Statelong1"
import {Statelong2} from "./Statelong2"
export class StateController
{
    private  _controllee: IControllee;
    private _currentState: BaseState | null;
    private _stateList: BaseState[];
    readonly InstanceOfProcessData: BaseState;
    readonly InstanceOfState1: BaseState;
    readonly InstanceOfState2: BaseState;
    readonly InstanceOfState3: BaseState;
    readonly InstanceOfState4: BaseState;
    readonly InstanceOflong1: BaseState;
    readonly InstanceOflong2: BaseState;
    readonly MaxNumberOfStateIDs: number = 7;
    public constructor(controllee: IControllee)
    {
        this._controllee = controllee;
        this.InstanceOfProcessData = new StateProcessData(this, this._controllee);
        this.InstanceOfState1 = new StateState1(this, this._controllee);
        this.InstanceOfState2 = new StateState2(this, this._controllee);
        this.InstanceOfState3 = new StateState3(this, this._controllee);
        this.InstanceOfState4 = new StateState4(this, this._controllee);
        this.InstanceOflong1 = new Statelong1(this, this._controllee);
        this.InstanceOflong2 = new Statelong2(this, this._controllee);
        this._currentState = this.InstanceOfState1;
        this._stateList = [
            this.InstanceOfProcessData,
            this.InstanceOfState1,
            this.InstanceOfState2,
            this.InstanceOfState3,
            this.InstanceOfState4,
            this.InstanceOflong1,
            this.InstanceOflong2
        ];
    }
    public GetCurrentIdFromStateId(parentStateId: number): number
    {
        if(parentStateId == -1)
        {
            return this.GetCurrentIdFromState(this._currentState);
        }
        if(parentStateId < -1 || parentStateId >= this.MaxNumberOfStateIDs)
        {
            return -1;
        }
        return this.GetCurrentIdFromState(this._stateList[parentStateId].CurrentSubState());
    }
    private GetCurrentIdFromState(state: BaseState | null): number
    {
        if(state == null)
        {
            return -1;
        }
    return state.GetStateID();
    }
    public ResumeState(parentStateId: number, stateId: number): void
    {
        let state: BaseState | null = null;
        if(stateId >= 0 && stateId < this.MaxNumberOfStateIDs)
        {
            state = this._stateList[stateId];
        }
        if(parentStateId == -1)
        {
            this._currentState = state;
            return;
        }
        if(parentStateId < -1 || parentStateId >= this.MaxNumberOfStateIDs)
        {
            return;
        }
        this._stateList[parentStateId].ResumeSubState(state);
    }
    public StateName(parentStateId: number): string
    {
        if(parentStateId >= 0 && parentStateId < this.MaxNumberOfStateIDs)
        {
            return this._stateList[parentStateId].GetStateName();
        }
        return "(None)";
    }
    public TryTransitWithoutEvent(): boolean
    {
        if(this._currentState == null)
        {
            return false;
        }
        let current: BaseState | null = this._currentState;
        this._currentState = this._currentState.TryTransitWithoutEvent();
        if(this._currentState != null)
        {
            this._currentState = this._currentState.OutlineState();
        }
        return (current != this._currentState);
    }
    public TransitAborted(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitAborted();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("Aborted");
        }
    }
    public TransitDeepResume(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitDeepResume();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("DeepResume");
        }
    }
    public TransitEnoughData(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitEnoughData();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("EnoughData");
        }
    }
    public TransitFailed(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitFailed();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("Failed");
        }
    }
    public TransitNewData(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitNewData();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("NewData");
        }
    }
    public TransitPause(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitPause();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("Pause");
        }
    }
    public TransitResume(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitResume();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("Resume");
        }
    }
    public TransitSucceeded(): void
    {
        if(this._currentState != null)
        {
            this._currentState = this._currentState.TransitSucceeded();
            if(this._currentState != null)
            {
                this._currentState = this._currentState.OutlineState();
            }
        } else {
            this._controllee.OverTransition("Succeeded");
        }
    }
    public GetCurrentStateName(): string
    {
        if(this._currentState == null)
        {
            return "(end)";
        }
        else
        {
            return this._currentState.GetStateName();
        }
    }
}